glotaran/io/register.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import os
+
+import xarray as xr
+
+from glotaran.model import Model
+from glotaran.parameter import ParameterGroup
+from glotaran.project import Result
+from glotaran.project import SavingOptions
+from glotaran.project import Scheme
+
+from .io import Io
+
+_io_register = {}
+
+
+def register_io(fmt: str | list[str], io: Io):
+    fmt = fmt if isinstance(fmt, list) else [fmt]
+    for fmt_name in fmt:
+        _io_register[fmt_name] = io()
+
+
+def known_fmt(fmt: str) -> bool:
+    return fmt in _io_register
+
+
+def get_io(fmt: str):
+    if not known_fmt(fmt):
+        raise ValueError(f"Unknown format '{fmt}'. Known formats are: {known_fmts()}")
+    return _io_register[fmt]
+
+
+def known_fmts() -> list[str]:
+    return [fmt for fmt in _io_register]
+
+
+def load_model(file_name: str, fmt: str = None) -> Model:
+    fmt = _get_fmt_from_file_name(file_name) if fmt is None else fmt
+    io = get_io(fmt)
+    try:
+        return io.read_model(fmt, file_name)
```

Comment:
`fmt` and `file_name` should be the same order for both functions or called as keyword arguments.

Also, the way this is implemented there is no way pass additional `kwargs` to the implementation function.

Reply:
Nope, the a file loader can in prinziple support more then one extion and must know which one is requested. You are write about the order of arguments, I realized it only when there was already tons of implementation and I don't feel like refactoring all this. Can we live it it?

Reply:
I would prefer to have it refactored to be consistent, but it can be picked in a later maintenance round if you feel it's too much hassle to do it within this PR, which has gotten rather bulky already.

Reply:
Note that this is mostly important for reading the data since those readers can have plenty of options.
Tackling multiple supported formats is why I said that initializing `Io` is useful in this case.

Here is a rough idea how it could work:

```python
def get_io(fmt: str) -> Io:
    if not known_fmt(fmt):
        raise ValueError(f"Unknown format '{fmt}'. Known formats are: {known_fmts()}")
    return _io_register[fmt](fmt)

class Io:
    def __init__(self, fmt: str):
        # the register decorated can sanity check those values
        self.fmt = fmt

    def read_model(self, filename: str | os.PathLike, **kwargs):
        if self.fmt == "fmt1":
            ...

    def read_dataset(self, filename: str | os.PathLike, *, option1=1, option1=1):
        if self.fmt == "fmt1":
            ...

    def get_data_loader(self, fmt: str):
        return self.read_dataset

```

Reply:
Since `register_io` already returns instances having `self.fmt` would make `fmt` in the methods obsolete (simpler API).